Cookie Store API docs updates - live examples

hamishwillee opened the pull request:
FF134 adds support for a (large) subset of the Cookie Store API.

The docs are good, but perhaps a little "thin" on examples. I created some live examples in order to test my understanding: this PR adds them to the docs. 
Note that they are not perfect - it is hard to control the domain etc in live examples, and hence demonstrate some of the API options.

Related docs work can be tracked in #37929

hamishwillee committed:
CookieStore.set() live examples

hamishwillee committed:
CookieStore.get() live example

hamishwillee committed:
CookieStore.getAll() live example

hamishwillee committed:
CookieStore.delete() live example

hamishwillee committed:
CookieStore exmaples

hamishwillee commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
+
+Press **Show cookies** to set the cookie and then display it.
+Note that some browsers will only display the `name` and `value`, while others will display all the properties of the cookie.
+Even if the values are not displayed, they are still set.
```
@wbamberg FYI
FF has implemented this so that if you set the values they are sent internally, but all getters and events only return name/value. 

While I have updated the getter/event BCD to indicate this, it isn't clear in this example because it is a setter() and that part of the BCD isn't shown here.

wbamberg commented on files/en-us/web/api/cookiestore/getall/index.md:
```diff
@@ -92,18 +93,103 @@ Each object contains the following properties:
-} else {
-  console.log("Cookie not found");
+We then await on `getAll()`, without specifying any parameters.
+If the returned Promise resolves with an object we iterate the array, logging the properties of each cookie.
```
```suggestion
If the returned promise resolves with an object we iterate the array, logging the properties of each cookie.
```

wbamberg commented on files/en-us/web/api/cookiestore/getall/index.md:
```diff
@@ -92,18 +93,103 @@ Each object contains the following properties:
+Press **Show cookies** to add and then get all the cookies.
+Note that some browsers will only display the `name` and `value`, while others will display all the properties of each cookie.
+
+{{EmbedLiveSample('Get all cookies for this context', 100, 350)}}
```
I get a scrollbar for this, need a bigger iframe?

wbamberg commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
-The following example sets a cookie by passing an object with `name`, `value`, `expires`, and `domain`.
+### Setting a cookie with name and value
+
+The following example sets a cookie by passing a `name` and `value` of cookie1 and cookie1-value, respectively.
```
```suggestion
The following example sets a cookie by passing a `name` and `value` of "cookie1" and "cookie1-value", respectively.
```

wbamberg commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
+
+#### JavaScript
+
+The example code first awaits for the cookie to be set, then waits to get the value and log it (note the logging and other display code is omitted).
```
```suggestion
The example code first waits for the cookie to be set, then waits to get the value, then logs the value (note that the logging and other display code is omitted).
```

wbamberg commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
+#### JavaScript
+
+The example code first awaits for the cookie to be set, then waits to get the value and log it (note the logging and other display code is omitted).
+Note that asynchronous code is called within an `async` function because you can't `await` on a top level function.
```
I don't think you need this line.

hamishwillee replied:
```suggestion
```

wbamberg commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
+
+### Setting a cookie with options
+
+The following example sets a cookie by passing an "options" object with `name`, `value`, `expires`, and `partitioned`.
```
```suggestion
The following example sets a cookie by passing an `options` object with `name`, `value`, `expires`, and `partitioned`.
```

wbamberg commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +69,172 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
 ## Examples
 
-The following example sets a cookie by passing an object with `name`, `value`, `expires`, and `domain`.
+### Setting a cookie with name and value
```
I can't comment on ^, but we should define SameSite in one place and link to it, rather than have loads of half-correct definitions of it in various places.

hamishwillee replied:
We do define SameSite in the `Set-Cookie` document which is linked. There is one fine-point difference that this API explicitly defines the enum values, so hence lower case.

I take you point on the duplication, and I don't think the description is necessary. So for all cases in this doc I have replaced with text like :

> One of the following [`SameSite`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value) values: [`"strict"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#strict), [`"lax"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#lax), or [`"none"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#none).



https://github.com/mdn/content/pull/38095/commits/9f81fa84a0ceaa946abedb480371831520f612be

wbamberg replied:
Yeah, I'm a bit sensitive to this at the moment because I found multiple definitions of this while researching CSRF, and most of them are incomplete.

- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#controlling_third-party_cookies_with_samesite
- https://developer.mozilla.org/en-US/docs/Web/Security/Practical_implementation_guides/CSRF_prevention#samesite_cookies

hamishwillee replied:
Fair enough. They do have slightly different audiences so the content probably wouldn't quite match, but they should at least link to a definitive meaning as per HTTP in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
-In this example, we set a cookie and write to the console feedback as to whether the operation succeeded or failed.
+### Setting cookies
+
+The following example sets cookie by passing a `name` and `value` and then by setting an options value.
```
```suggestion
The following example sets cookies by passing a `name` and `value` and then by setting an `options` value.
```

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+#### JavaScript
+
+The example code sets two cookies, the first using `name` and `value` parameters, and the second using an "options" object with `name`, `value`, `expires`, and `partitioned` properties set.
```
```suggestion
The example code sets two cookies, the first using `name` and `value` parameters, and the second using an `options` object with `name`, `value`, `expires`, and `partitioned` properties set.
```

wbamberg replied:
But this is a weird way to put it. They both use an `options` object. I wouldn't say that the first uses "`name` and `value` parameters".

hamishwillee replied:
Yes it is. Fixed in https://github.com/mdn/content/pull/38095/commits/3fadf3abceca2ce94f5c263d64b0753a1fc0008d

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+The example code sets two cookies, the first using `name` and `value` parameters, and the second using an "options" object with `name`, `value`, `expires`, and `partitioned` properties set.
+
+The {{apiref("CookieStore.get()")}} method is then used to get the cookies, which are then logged.
```
```suggestion
The {{domxref("CookieStore.get()")}} method is then used to get the cookies, which are then logged.
```

(but you could make it less passive with "We then use...")

hamishwillee replied:
FYI, I modified your suggestion to be less passive.

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+#### JavaScript
+
+The example code first sets a number of cookies that we'll use for demonstrating the get methods.
```
```suggestion
The example code first sets three cookies that we'll use for demonstrating the get methods.
```

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+The example code first sets a number of cookies that we'll use for demonstrating the get methods.
+
+First it creates `cookie1` and `cookie2` using the {{domxref("CookieStore.set()")}} method.
+Then it creates a couple of using the older synchronous {{domxref("Document.cookie")}} property (just so we can show that these are also fetched using our get methods)
```
```suggestion
Then it creates a third cookie using the older synchronous {{domxref("Document.cookie")}} property (just so we can show that these are also fetched using the `get()` and `getAll()` methods).
```

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+### Delete a named cookie
+
+In this example, a cookie is deleted by passing its name to the `delete()` method.
```
```suggestion
In this example, we delete a cookie by passing its name to the {{domxref("CookieStore.delete()","delete()")}} method.
```

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+The code first sets two cookies (which we then use to demonstrate deletion).
+
+We then list the names of both cookies (code for getting the cookie names not shown), delete one of the cookies, and then list all cooking names again.
```
```suggestion
We then list the names of both cookies (code for getting the cookie names not shown), delete one of the cookies, and then list all cookie names again.
```

hamishwillee replied:
I guess I was a bit hungry when I wrote this?

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +31,304 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+Press **Show cookies** to run the code above.
+This should show that the deleted cookie is present in the first list, and not in the second.
+
+{{EmbedLiveSample('Delete a named cookie', 100, 190)}}
```
Hm, interesting, for me the `favorite_food` cookie from the last example is still here in this one!

hamishwillee replied:
Indeed. If you run the examples one after the other it will be present. However if you reset an example they are gone.

Do I need to fix this? I'm going to assume not - based on the fact that it still demonstrates deletion of the indicated cookie.

wbamberg replied:
I think it is worth fixing, because it's confusing otherwise. As a reader I expect examples to be self-contained. I think it would be OK to have some delete-cookies code you run at the end, that you could even hide.

hamishwillee replied:
@wbamberg Unfortunately I've run into problems, for which I need your advice.

1. First I tried to add this code - to do so I created some deletion code in https://github.com/mdn/content/pull/38095/commits/f3246bc2b4a6af4d7fa1f300d83c662791bc849d . 
   - When tested _locally_ this indicates that a cookie created using the `document.cookie` is not deleted. 
   - You can test this by doing the same test you that cause this discussion (`Getting cookies` before `Delete a named cookie` example 
   - the log code shows that cookie deletion succeeds for all the cookies - the method should throw if it can't delete.
   - BUT ... when I read the cookies it is still there.
   - Note that I left the logging code present just to demonstrate this.
   - Note, I use the word locally, because when I try on the PR such as [delete_a_named_cookie](https://pr38095.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CookieStore#delete_a_named_cookie) nothing is displayed - see next point.
2. When I try on the PR the examples are empty or indicate that the cookie does not exist.
   - I updated the `set()` example  [setting_a_cookie_with_options](https://pr38095.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CookieStore/set#setting_a_cookie_with_options)  to log the creation of the cookie.
   - On Chrome this gives and error `Error setting cookie2: UnknownError: Failed to execute 'set' on 'CookieStore': An unknown error occurred while writing the cookie.`
   - On FF NIghtly no error is raised - the set succeeds, but the cookie then is not read.
3. ON play for these examples the output is empty on Chrome and Gives me 404 text on  FF nightly.
4. Local test file used instead. This works if served from localhost:

   ```html
     <!DOCTYPE html>
      <html>
        <head>
          <meta charset="utf-8" />
          <title>Test CookieStore.set()</title>
          <style>
            #log {
              height: 180px;
              overflow: scroll;
              padding: 0.5rem;
              border: 1px solid black;
            }
      
            div:focus {
              border: 2px solid red;
            }
          </style>
        </head>
        <body>
          <h1>Test CookieStore.set()</h1>
      
          <button id="showCookies" type="button">Show cookies</button>
          <button id="reset" type="button">Reset</button>
          <pre id="log"></pre>
      
          <script>
            const reload = document.querySelector("#reset");
      
            reload.addEventListener("click", () => {
              window.location.reload(true);
            });
      
            const logElement = document.querySelector("#log");
      
            function log(text) {
              logElement.innerText = `${logElement.innerText}${text}\n`;
              logElement.scrollTop = logElement.scrollHeight;
            }
      
            function logCookie(name, cookie) {
              if (cookie) {
                log(`${name}:`);
                for (const [key, value] of Object.entries(cookie)) {
                  log(` ${key}: ${value}`);
                }
              } else {
                log(`${name}: Cookie not found`);
              }
            }
      
            async function cookieTest() {
              const day = 24 * 60 * 60 * 1000;
      
              try {
                // Set cookie
                await cookieStore.set({
                  name: "cookie2",
                  value: `cookie2-value`,
                  expires: Date.now() + day,
                  partitioned: true,
                });
              } catch (error) {
                log(`Error setting cookie2: ${error}`);
                console.log(error);
              }
      
              // Log the new cookie
              const cookie = await cookieStore.get("cookie2");
              logCookie("cookie2", cookie);
            }
      
            const showCookies = document.querySelector("#showCookies");
      
            showCookies.addEventListener("click", () => {
              cookieTest();
            });
          </script>
        </body>
      </html>
   ```
   - Note this file above fails with specific error if just loaded from file system, but only on Chrome (`Error setting cookie2: TypeError: Failed to execute 'set' on 'CookieStore': Cannot modify a secure cookie on insecure origin`) - on FF just doesn't work
   
I know that's a lot to digest. It seems to me that the unknown error is causing the problems. I suspect it is something to do with the PR server environment, but can't think what to try. Ideas?

wbamberg replied:
I will look at this properly tomorrow, but just a note: none of these examples worked for me in the PR preview (and I saw the same errors you see), so I checked out your fork and ran it locally, and it all worked there. I don't know why (and probably should have mentioned it but forgot). Which I think is your point (2) right?

The page for [document.cookie](https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie#notes) says:

> You can delete a cookie by updating its expiration time to zero.

Have you tried that?

hamishwillee replied:
@wbamberg Regarding point 2 - yes "doesn't work in PRs and FF doesn't provide any debug about that".

I haven't tried deleting the cookie by updating its expiration time to zero. Kind of glad of that, because I think there are bugs here on deletion implementation. (I mean, yes I can do this, but we'll still be stuck.

Thanks for your help!

wbamberg replied:
I don't have any answers, but I have more confirmation that something is weird. I wrote a little page to test: https://github.com/wbamberg/cookie-test. You can:

- add cookies, either using the cookie store or using document.cookie
- delete a cookie, delete all cookies, or list all cookies

If I run this locally using python's http.server, it works fine. If I run it from https://wbamberg.github.io/cookie-test/, I see exactly what you do: I can't delete cookies that are set using document.cookie.

I don't have any explanation for this yet.

hamishwillee replied:
THanks very much @wbamberg - that's helpful. To me a clear indication that there are at least some bugs here. Enough to make it worth asking for an explanation in the FF implementation issue https://bugzilla.mozilla.org/show_bug.cgi?id=1937477#c14

Let's reconsider this later in the week.

hamishwillee replied:
@wbamberg FYI the cookies created with  `document.cookies` are not being deleted because they don't have a matching path. This is because that method using the path of the path the cookie is created in, by default, while `set()` always uses `/` by default, and `delete()` matches against this.

This feels very painful because Firefox have chosen not to return values for cookies other than name and value in the `get()` and `getAll()` - so you can't reliably query for the paths to delete these. All you can do is either store the values or more likely always set the default if you have to use the document.cookies for whatever reasons. I guess it isn't really a problem if you're the one setting the code - you just won't use that API anymore.

The other problem is that there is no error if the cookie doesn't match. I'm checking that. 

Still working on why this doesn't work on the live site.

wbamberg reviewed (changes requested):
nice examples Hamish!

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -52,14 +52,99 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when deletion
+
+#### JavaScript
+
+The code first sets two cookies (which we then use to demonstrate deletion).
+
```
```suggestion
The code first sets two cookies.
```

hamishwillee commented on files/en-us/web/api/cookiestore/get/index.md:
```diff
@@ -93,18 +95,87 @@ A {{jsxref("Promise")}} that resolves with an object representing the first cook
+
+#### JavaScript
+
+The code first sets a cookie (which we then use to demonstrate `get()`).
```
It's intentional, but I agree pretty pointless.
```suggestion
The code first sets a cookie.
```

hamishwillee committed:
add no-param getAll() to syntax

hamishwillee committed:
Apply easy suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
````diff
@@ -52,14 +52,99 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when deletion
+
+```js hidden
+// List cookies
+async function getCookieNames() {
+  let names = "";
+  const cookies = await cookieStore.getAll();
+  if (cookies.length > 0) {
+    cookies.forEach((cookie) => (names += `${cookie.name} `));
+  }
+  return names;
+}
````
I hid it as "not sufficiently relevant". I'll add in the one line solution because perhaps it is interesting :-)

hamishwillee committed:
delete - add getall cookie lines to example

hamishwillee committed:
Fix getAll() example frame size

hamishwillee committed:
CookieStore - example wording for options

hamishwillee committed:
Update files/en-us/web/api/cookiestore/set/index.md

hamishwillee committed:
Link to SameSite for values

hamishwillee commented:
Thanks @wbamberg - I have addressed the issues I think were needed. Let me know if good to go.

hamishwillee committed:
Log code shows deletion doesn't always work

hamishwillee committed:
Set - log errors

hamishwillee commented on files/en-us/web/api/cookiestore/set/index.md:
```diff
@@ -69,19 +62,180 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when setting
+    });
+  } catch (error) {
+    log(`Error setting cookie2: ${error}`);
+    console.log(error);
```
Not, this will have to be removed.

hamishwillee committed:
Fix up all the methods

hamishwillee committed:
Delete doesn't fail on mismatch

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -48,18 +49,392 @@ A {{jsxref("Promise")}} that resolves with {{jsxref("undefined")}} when deletion
   - : Thrown if the origin can not be {{glossary("Serialization", "serialized")}} to a URL.
 - {{jsxref("TypeError")}}
-  - : Thrown if deleting the cookie represented by the given `name` or `options` fails.
+  - : Thrown if a cookie that matches a given `name` or `options` cannot be deleted.
```
Note the change. The previous way could imply that the method should fail if you don't match the passed options. But actually I am "pretty sure" that if there is no match it silently fails.  I will have to revisit this doc a bit if I can confirm that.

hamishwillee replied:
It does. Updated.

hamishwillee committed:
Water down the note on set() params

hamishwillee committed:
Fix up the CookieStore top page

hamishwillee committed:
Fix getter example

hamishwillee committed:
Disclaimer on examples

hamishwillee commented:
@wbamberg 

1. The problem with deleting cookies created with `document.cookie` was they the get the path of the current page, whereas cookies created with the `CookieStore.set()` get the path `/`. So I was attempting to delete a non-matching cookies.

   I have fixed this where needed. I also updated the docs for `delete()` to make it clear that not matching fails silently.
   
2. The examples fail because cookies cannot be set in the MDN environment - this is also true for `document.cookie` (its examples don't work). I tried as a new interactive example too. Since the error is "unknown" (and not reported at all in FF) there is nothing we can really do to fix this. 

   Now I could have removed the live example and just left my code, but I am kind of hoping they will eventually give us a better error so this might be fixed. I also kind of like that you can run this locally if you want, and see all the code by running in play.

  So what I have done is added a warning before the example sections:

   > [!WARNING]
   > These examples do not run properly within the MDN environment because setting cookies results in an unknown error.
   > The examples can be tested by copying the source code and running it with a [local server](https://developer.mozilla.org/en-US/docs/Learn_web_development/Howto/Tools_and_setup/set_up_a_local_testing_server), or by [building this documentation locally](https://github.com/mdn/content?tab=readme-ov-file#build-the-site).

   Is this acceptable? I might presumably also create a glitch with some of this or an example repo of some kind. But I'm hoping you say no, this is OK.

wbamberg commented:
> The examples fail because cookies cannot be set in the MDN environment

So am I understanding you right that none of these examples will work on the live site? If so I can't see any point in making them live samples at all. You could just modify the JS to use `console.log()` instead of `log()` and make them static samples, i.e.:

```js
async function setTestCookies() {
  // Set two cookies
  try {
    await cookieStore.set("cookie1", "cookie1-value");
  } catch (error) {
    console.log(`Error setting cookie1: ${error}`);
  }

  try {
    await cookieStore.set("cookie2", "cookie2-value");
  } catch (error) {
    console.log(`Error setting cookie2: ${error}`);
  }

  // Log cookie names
  const cookieNames = (await cookieStore.getAll())
    .map((cookie) => cookie.name)
    .join(" ");
    console.log(`Initial cookies: ${cookieNames}`);
}

async function cookieTest() {
  // Create our test cookies
  await setTestCookies();

  // Delete cookie1
  try {
    await cookieStore.delete("cookie1");
  } catch (error) {
    console.log(`Error deleting cookie1: ${error}`);
  }

  // Log cookie names again (to show cookie1 deleted)
  cookieNames = (await cookieStore.getAll())
    .map((cookie) => cookie.name)
    .join(" ");
    console.log(`Cookies after deleting cookie1: ${cookieNames}`);
}
```

...this code works for me in say https://example.org, and it seems easier to say to users "copy this code into your console and run it in example.org" than to say "here's a live sample that doesn't work, you have to run the site locally". I don't think anyone would do that.

hamishwillee committed:
Delete() no longer live code example:

hamishwillee committed:
get() no longer live code example:

hamishwillee committed:
getAll() no longer live code example:

hamishwillee committed:
set() no longer live code example:

hamishwillee committed:
Fix up examples in cookiestore

hamishwillee commented:
@wbamberg Fair enough. I didn't want to rewrite it after all the work doing the live examples - still hoping we'll get enough debug to get it working at some point. But it is pointless. Fixed, ready for rereview.

wbamberg commented:
>  I didn't want to rewrite it after all the work doing the live examples

Yes, I know, I'm sorry, and I'd feel the same way. But objectively, this looks like sunk cost fallacy.

hamishwillee commented:
> But objectively, this looks like sunk cost fallacy.

Yes :-(. Anyway, fixed up as per your suggestions. It is definitely better.

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
```diff
@@ -31,28 +35,147 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+
+This example shows how to set cookies by passing a `name` and `value`, and by setting an `options` value.
+
+The `cookieTest()` method first sets a cookie with `name` and `value` properties, while the second is set with `name`, `value`, and `expires` properties.
```
```suggestion
The `cookieTest()` method sets one cookie with `name` and `value` properties and another with `name`, `value`, and `expires` properties.
```

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
````diff
@@ -31,28 +35,147 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+cookieTest();
+```
+
+Note that some browsers will only display the `name` and `value`, while others will display all the properties of the cookie.
````
Again I'm not sure we need to call this out here.

wbamberg commented on files/en-us/web/api/cookiestore/index.md:
````diff
@@ -31,28 +35,147 @@ The `CookieStore` is accessed via attributes in the global scope in a {{domxref(
+}
 ```
 
+When run in a console we should see the deleted cookie ("cookie1") is present in the first log array, and not in the second.
````
Why "when run in a console" specifically?

hamishwillee replied:
Removed.

wbamberg commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+
+When run, the console log should show that both cookie1 and cookie2 are present initially, but cookie1 is not listed after it has been deleted.
+
+The code can be tested by copying it into a test harness and running it with a [local server](/en-US/docs/Learn_web_development/Howto/Tools_and_setup/set_up_a_local_testing_server), or deploying it to a website site such as GitHub pages.
```
Or just don't say anything? We don't usually describe this in these situations.

hamishwillee replied:
I'll modify to not say this anywhere/everywhere - as you say, this is expected.

wbamberg reviewed (commented):
Thanks for the updates Hamish.

Overall, I'm not sure we need all the "some browsers will only display the name and value" bits, as we have that in BCD. And not sure we need all the "The code can be tested by" - we never say that any other time we have non-live samples, and I ran them all in the console and they worked fine.

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+    .join(" ");
+  log(`Initial cookies: ${cookieNames}`);
+
+  // Delete doc_cookie1 (should fail)
```
Correct.
```suggestion
  // Delete doc_cookie1 (fails if cookie path is not the root ("/")
```

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+  try {
+    await cookieStore.delete("doc_cookie1");
+  } catch (error) {
+    log(`Error deleting doc_cookie1: ${error}`);
```
I don't know how I missed all these sorry. I even did a search. Worries me I might have missed pushing a commit.

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+  cookieNames = (await cookieStore.getAll())
+    .map((cookie) => cookie.name)
+    .join(" ");
+  log(`Cookies attempting deletion: ${cookieNames}`);
```
It was probably meant to be "Cookies after attempting deletion", but I like yours more.

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
````diff
@@ -41,25 +44,223 @@ Or
+```
+
+When run, the console log should show that both "doc_cookie1" and "doc_cookie2" are present initially.
+Afterwards `doc_cookie2` should not be present: it will have been deleted because its path matches the default path used by `delete()`.
+The cookie `doc_cookie2` will be removed if the code is run from a page in the root, but otherwise will not match the path, and hence not be removed.
````
OK, so this is a typo

```suggestion
Afterwards `doc_cookie2` should not be present: it will have been deleted because its path matches the default path used by `delete()`.
The cookie `doc_cookie1` will be removed if the code is run from a page in the root, but otherwise will not match the path, and hence not be removed.
```

I like your suggestion, though I was trying to show the paths might wrong by default. I'll merge this fix but I might well come back and do this as you suggested.

hamishwillee commented on files/en-us/web/api/cookiestore/get/index.md:
```diff
@@ -8,7 +8,7 @@ browser-compat: api.CookieStore.get
 {{securecontext_header}}{{APIRef("Cookie Store API")}}{{AvailableInWorkers("window_and_service")}}
 
-The **`get()`** method of the {{domxref("CookieStore")}} interface returns a single cookie with the given `name` or `options` object. The method will return the first matching cookie for the passed parameters.
+The **`get()`** method of the {{domxref("CookieStore")}} interface returns a {{jsxref("Promise")}} that resolves to a single cookie matching the given `name` or `options` object. The method will return the first cookie that matches.
```
Yes, because this only selects on name and URL, but you can have variants in different paths with the same name (though you can only return one found)

hamishwillee commented on files/en-us/web/api/cookiestore/getall/index.md:
````diff
@@ -92,18 +87,58 @@ Each object contains the following properties:
 ```
 
+When run in a browser this will display information about "cookie1" and "cookie2" in the console.
+Note that some browsers will only display the `name` and `value`, while others will display all the properties of the cookie.
````
You're right. I found this such a startling omission on firefox I wrote this everywhere. It is in BCD. Will remove.

hamishwillee replied:
Further, I thought it useful to add that the values can still be set, since that isn't entirely obvious. Anyway, removing.

hamishwillee committed:
Suggestions from code review - the easy ones

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Remove notes about returned properties and testing

hamishwillee committed:
remove another mention of console specifically

hamishwillee committed:
Tidy up the last delete example

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+
+This example shows how we can use `delete()` to remove a cookie created using {{domxref("document.cookie")}}.
+
+A potential issue is that cookies created with `document.cookie` have a default path equal to the path of the document they are created in (unlike cookies created with {{domxref("CookieStore.set()")}}, which have a default [`path`](#path) of `/`).
```
Yes. I link to that first in the next iteration.

hamishwillee committed:
Improve the document.cookies example

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,223 @@ Or
+This example shows how we can use `delete()` to remove a cookie created using {{domxref("document.cookie")}}.
+
+A potential issue is that cookies created with `document.cookie` have a default path equal to the path of the document they are created in (unlike cookies created with {{domxref("CookieStore.set()")}}, which have a default [`path`](#path) of `/`).
+In order to be able to know the path later when we match the cookie, we can explicitly set the value when creating the cookie.
```
Yes it does read oddly, and it is.

On reflection, providing a solution is pointless because how you fix the problem depends on how you are using your cookies. If you're using root only cookies then you might indeed always explicitly set them to root.
But if you're using path based cookies, then in both cases you'll need a way to work out what your cookie path is.

I have rewritten this to highlight the issue without this (or any) particular "solution".

hamishwillee commented on files/en-us/web/api/cookiestore/get/index.md:
```diff
@@ -64,14 +66,7 @@ A {{jsxref("Promise")}} that resolves with an object representing the first cook
-      - : Cookies are not sent on normal cross-site subrequests (for example to load images or frames into a third party site), but are sent when a user is navigating within the origin site (i.e. when following a link).
-    - `"none"`
-      - : Cookies will be sent in all contexts.
+  - : One of the following [`SameSite`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#samesitesamesite-value) values: [`"strict"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#strict), [`"lax"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#lax), or [`"none"`](/en-US/docs/Web/HTTP/Headers/Set-Cookie#none).
```
Gah, I didn't write that. This comes from https://wicg.github.io/cookie-store/#dom-cookiestore-get-options :

> If options["[url](https://wicg.github.io/cookie-store/#dom-cookiestoregetoptions-url)"] is present, then run these steps:
> 1. Let parsed be the result of [parsing](https://url.spec.whatwg.org/#concept-basic-url-parser) options["[url](https://wicg.github.io/cookie-store/#dom-cookiestoregetoptions-url)"] with settings’s [API base URL](https://html.spec.whatwg.org/multipage/webappapis.html#api-base-url).
> 1. If [this](https://webidl.spec.whatwg.org/#this)'s [relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global) is a [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window) object and parsed does not [equal](https://url.spec.whatwg.org/#concept-url-equals) url, then return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).
> 1. If parsed’s [origin](https://url.spec.whatwg.org/#concept-url-origin) and url’s [origin](https://url.spec.whatwg.org/#concept-url-origin) are not the [same origin](https://html.spec.whatwg.org/multipage/browsers.html#same-origin), then return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).
> 1. Set url to parsed.

I read this as the creation URL is the URL of the origin that created the cookie. If it was a window they need to be precisely equal - including path, while if created by something else (I presume a worker) it just has to match the resource origin - so no path.

Does that make sense?

hamishwillee committed:
Fix up get creation url query

hamishwillee committed:
Specify the TypeError reasons in get and getAll

hamishwillee commented on files/en-us/web/api/cookiestore/get/index.md:
```diff
@@ -87,22 +82,42 @@ A {{jsxref("Promise")}} that resolves with an object representing the first cook
     - The `options` parameter is an empty object.
-    - The `url` option is present and is not equal with the creation URL, if in main thread.
-    - The `url` option is present and its origin is not the same as the origin of the creation URL.
+    - The method is called in the main thread, and the `url` option is specified but does not match the URL of the current window.
+    - The method is called in a worker and the `url` option is specified, but does not match the origin of the worker.
```
Continuing https://github.com/mdn/content/pull/38095/files#r1980081173

I think this is completely wrong. The creationURL is mentioned only in the context where you don't specify the `url` option.

The cases where `url` matters come from https://wicg.github.io/cookie-store/#dom-cookiestore-get-options :

> If options["[url](https://wicg.github.io/cookie-store/#dom-cookiestoregetoptions-url)"] is present, then run these steps:
> 1. Let parsed be the result of [parsing](https://url.spec.whatwg.org/#concept-basic-url-parser) options["[url](https://wicg.github.io/cookie-store/#dom-cookiestoregetoptions-url)"] with settings’s [API base URL](https://html.spec.whatwg.org/multipage/webappapis.html#api-base-url).
> 1. If [this](https://webidl.spec.whatwg.org/#this)'s [relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global) is a [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window) object and parsed does not [equal](https://url.spec.whatwg.org/#concept-url-equals) url, then return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).
> 1. If parsed’s [origin](https://url.spec.whatwg.org/#concept-url-origin) and url’s [origin](https://url.spec.whatwg.org/#concept-url-origin) are not the [same origin](https://html.spec.whatwg.org/multipage/browsers.html#same-origin), then return [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).
> 1. Set url to parsed.


I read "2" as an error if a main window tries to get a cookie at some other URL. YOu can only access cookies this page has a right to see. 

I read "3" as if  the same thing but you're calling the method from a worker. IN this case the test only looks at the origin. 

Does that make sense?

I have rolled this fix to getAll as well.

wbamberg replied:
I think so? I think from the note above that all this is to enable a service worker to get any cookie under its scope. I'm not sure why this bit of the spec talks about origins instead. Are they the same thing for service workers?

hamishwillee replied:
I don't _know_ either. But what it now says matches the spec, and what was there is wrong. So I'm pretty happy this is "less wrong".

wbamberg commented on files/en-us/web/api/cookiestore/delete/index.md:
````diff
@@ -41,25 +44,223 @@ Or
+cookieTest();
 ```
 
+When run, the console log should show that both "cookie1" and "cookie2" are present initially, but "cookie2" is not listed afterwards.
+The cookie named "cookie1" is still present because it does not match the cookies specified in the `delete()` call.
````
@hamishwillee , in Chrome I see:

> Cookies remaining after deleting cookie1: cookie1

...which I think is what the text was expecting:

> When run, the console log should show that both "cookie1" and "cookie2" are present initially, but "cookie2" is not listed afterwards. The cookie named "cookie1" is still present because it does not match the cookies specified in the delete() call.

...but the "Cookies remaining after deleting cookie1" then reads weirdly of course. Perhaps say instead "Cookies remaining after attempted deletions" instead?

BUT in Firefox I get:

> Cookies remaining after deleting cookie1: cookie2

...which is not what I was expecting.

hamishwillee replied:
I get the same result: 
- FF: Cookies remaining after deleting cookie1: cookie2
- Chrome: Cookies remaining after deleting cookie1: cookie1

Since both cookies are created with `partitioned: true` and only the `delete()` call for cookie2 specifies this property, only cookie2 should be deleted. That means FF is deleting the wrong cookie, and as you say the text is wrong for what we expect. Doh.

I have changed the text to `Cookies remaining after attempted deletions (cookie2 should be deleted): cookie2`
Now we get the following, which will is accurate, and highlights that FF does the wrong thing.

- FF: Cookies remaining after attempted deletions (cookie2 should be deleted): cookie2
- Chrome: Cookies remaining after attempted deletions (cookie2 should be deleted): cookie1

I'll ask about the FF error.

hamishwillee replied:
https://bugzilla.mozilla.org/show_bug.cgi?id=1937477#c18

wbamberg commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,213 @@ Or
+  }
+
+  // Log cookie names again (to show cookie1 deleted)
+  cookieNames = (await cookieStore.getAll())
```
```suggestion
  const cookieNames = (await cookieStore.getAll())
```

wbamberg commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -41,25 +44,213 @@ Or
+
+When run, the first log should show that both cookies are present.
+The second log should not include "doc_cookie2", as it should have matched and been deleted.
+It should include "doc_cookie1" because `/some_path` will not match the default deletion path (`/`).
```
When I run this I get:

> Initial cookies: doc_cookie2
> Cookies remaining: 

...it looks as if setting doc_cookie1 failed, presumably because it wanted the path to match the current page.

hamishwillee replied:
Yes. As per other discussion, setting a path fails unless you put the `'path'` as though it were a string.

wbamberg reviewed (commented):
Just a couple of comments although I am seeing some cross-browser weirdness in one of the delete() examples: https://github.com/mdn/content/pull/38095#discussion_r1992636915.

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Fix up deletion code to note that deletion might fail

hamishwillee commented on files/en-us/web/api/cookiestore/delete/index.md:
```diff
@@ -211,11 +215,11 @@ The code then deletes both cookies without specifying a `path` match option, and
   // Create doc_cookie1 with path /some_path
   document.cookie =
-    "doc_cookie1=doc_cookie1_name; SameSite=None; Secure; path=/some_path";
+    "doc_cookie1=doc_cookie1_name; SameSite=None; Secure; max-age=10; path='/some_path'";
```
FYI the need for ' around the path took me about an hour of trial and error to discover. I haven't updated the `document.cookie` though I should. 

The max-age setting deletes this cookies after 10 seconds, which we can do in this case because all the code is in one method. What I'd like to do is apply this everywhere so that we self-clean our cookies - but until this works in the MDN  environment, writing the examples so this works would be "clutter". It is OK here.

hamishwillee commented:
OK, I think it is fixed up ... but then I thought that last time. There is still an incompatibility - as linked in the open comment https://github.com/mdn/content/pull/38095#discussion_r1994693560 (and chasing up here https://bugzilla.mozilla.org/show_bug.cgi?id=1937477#c18).

wbamberg reviewed (approved):
👍 thanks for all your work on this Hamish!
